lisa/sut_orchestrator/hyperv/platform_.py

```diff
@@ -0,0 +1,295 @@
+
+from .. import HYPERV
+from .context import NodeContext, get_node_context
+from .schema import HypervNodeSchema, HypervPlatformSchema
+from .serial_console import SerialConsole, SerialConsoleLogger
+
+
+class _HostCapabilities:
+    def __init__(self) -> None:
+        self.core_count = 0
+        self.free_memory_kib = 0
+
+
+class HypervPlatform(Platform):
+    def __init__(
+        self,
+        runbook: schema.Platform,
+    ) -> None:
+        super().__init__(runbook=runbook)
+
+    @classmethod
+    def type_name(cls) -> str:
+        return HYPERV
+
+    @classmethod
+    def supported_features(cls) -> List[Type[feature.Feature]]:
+        return [SerialConsole]
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        hyperv_runbook = self.runbook.get_extended_runbook(HypervPlatformSchema)
+        assert hyperv_runbook, "platform runbook cannot be empty"
+        self._hyperv_runbook = hyperv_runbook
+        print("Hyperv platform initialize")
+        print(f"{self._hyperv_runbook}")
+
+        if len(self._hyperv_runbook.servers) > 1:
+            self._log.warning(
+                "Multiple servers are currently not supported. "
+                "Only the server host will be used."
+            )
+
+        server = self._hyperv_runbook.servers[0]
+        self.server_node = RemoteNode(
+            runbook=schema.Node(name="hyperv-server"),
+            index=-1,
+            logger_name="hyperv-server",
+            parent_logger=get_logger("hyperv-platform"),
+        )
+        self.server_node.set_connection_info(
+            address=server.address, username=server.username, password=server.password
+        )
+
+    def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
+        return self._configure_node_capabilities(environment, log)
+
+    def _configure_node_capabilities(
+        self, environment: Environment, log: Logger
+    ) -> bool:
+        if not environment.runbook.nodes_requirement:
+            return True
+
+        host_capabilities = self._get_host_capabilities(log)
+        nodes_capabilities = self._create_node_capabilities(host_capabilities)
+
+        nodes_requirement = []
+        for node_space in environment.runbook.nodes_requirement:
+            if not node_space.check(nodes_capabilities):
+                return False
+
+            requirement = node_space.generate_min_capability(nodes_capabilities)
+            nodes_requirement.append(requirement)
+
+        environment.runbook.nodes_requirement = nodes_requirement
+        return True
+
+    def _get_host_capabilities(self, log: Logger) -> _HostCapabilities:
+        host_cap = _HostCapabilities()
+
+        free_mem_bytes = self.server_node.tools[PowerShell].run_cmdlet(
+            "(Get-CimInstance -ClassName Win32_OperatingSystem).FreePhysicalMemory"
+        )
+        host_cap.free_memory_kib = int(free_mem_bytes) // 1024
+
+        lp_count = self.server_node.tools[PowerShell].run_cmdlet(
```

Comment:
Host information can be cached to prevent loading when VM created every time.

Reply:
Done.